py-pyside2: add v5.15.10

melven opened the pull request:
Adds most recent pyside2 version for qt5 plus mor dependency information concerning supported python versions.

melven committed:
py-pyside2: add version 5.15.10 and dep info

adamjstewart commented on var/spack/repos/builtin/packages/py-pyside2/package.py:
```diff
@@ -17,6 +17,7 @@ class PyPyside2(PythonPackage):
     # https://wiki.qt.io/Qt_for_Python_Development_Getting_Started
 
     version("develop", tag="dev")
+    version("5.15.10", tag="v5.15.10-lts-lgpl", submodules=True)
```
Do you know why this version isn't available on PyPI? I was going to say we should just use PyPI instead of git for downloading

melven replied:
I don't know, I just looked for a newer version because I needed it to fulfil version constraints (something else depended on newer qt in my toolchain).
The documentation on available versions of pyside2 for qt5 did not look straight-forward to me...

adamjstewart commented on var/spack/repos/builtin/packages/py-pyside2/package.py:
```diff
@@ -30,8 +31,12 @@ class PyPyside2(PythonPackage):
-    depends_on("python@2.7.0:2.7,3.5.0:3.5,3.6.1:", type=("build", "run"))
-    depends_on("python@2.7.0:2.7,3.5.0:3.5,3.6.1:3.8", when="@:5.14", type=("build", "run"))
+    # dependency matrix from: https://wiki.qt.io/Qt_for_Python
+    depends_on("python@2.7.0:2.7,3.5.0:3.5,3.6.1:3.8", when="@:5.15.0", type=("build", "run"))
+    depends_on(
+        "python@2.7.0:2.7,3.5.0:3.5,3.6.1:3.9", when="@5.15.1:5.15.7", type=("build", "run")
+    )
+    depends_on("python@3.5:3.10", when="@5.15.9:5.15.10", type=("build", "run"))
```
Spack no longer supports Python 3.6 and older so you could simplify this by saying:
```python
depends_on("python@:3.10", when="@5.15.8:", type=("build", "run"))
depends_on("python@:3.9", when="@5.15.1:5.15.7", type=("build", "run"))
depends_on("python@:3.8", when="@:5.15.0", type=("build", "run"))
```

adamjstewart reviewed (commented):
Unclear where to find deps for this package. Also adding @manuelakuhn since she recently worked on py-pyside

manuelakuhn reviewed (commented):
The requirements are confusing. I took a look in the source you can get by `git clone` (or easier, check out `spack-src`, if the build fails as it did for me) and found
- `setup.py`, importing `build_scripts/setup_runner.py`, getting the config from `build_scripts.config`
- `README.pyside2.md`:
    >> PySide2 versions following 5.12 use a C++ parser based on [Clang](http://clang.org/). The Clang library (C-bindings), version 6.0 or higher is required for building. Prebuilt versions of it can be downloaded from [download.qt.io](http://download.qt.io/development_releases/prebuilt/libclang/).
- `requirements.txt`
	```
	numpy==1.16.6; python_version < '3'
	# numpy; python_version >= '3', installed in coin_*_instructions.py atm
	# see https://tinyurl.com/y3dm3h86 for details
	# setuptools from v45+ removed the support for Python 2, so we pin an old release
	setuptools==44.0.0; python_version < '3'
	setuptools; python_version >= '3'
	sphinx
	six
	wheel>=0.35
	PyOpenGL
	pyinstaller==3.6
	```

`py-pyside2` does not build for me with the error
```
     1850    (quickcontrols2) Clang: 1 diagnostic messages:
  >> 1851    $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/qt-5.15.10-tzxivbcuo
             olou25ch2vpomclancshqzz/include/QtGui/qopengl.h:141:13: fatal: 'GL/gl.h' file not found
```
I am a little confused why there is a error message about clang

melven committed:
py-pyside2: simplify python versions constraints

melven committed:
py-pyside2: add +clang in llvm dependency

From README.pyside2.md:
  PySide2 versions following 5.12 use a C++ parser based on Clang. The
  Clang library (C-bindings), version 6.0 or higher is required for
  building. Prebuilt versions of it can be downloaded from download.qt.io.

melven commented:
> `py-pyside2` does not build for me with the error
> 
> ```
>      1850    (quickcontrols2) Clang: 1 diagnostic messages:
>   >> 1851    $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/qt-5.15.10-tzxivbcuo
>              olou25ch2vpomclancshqzz/include/QtGui/qopengl.h:141:13: fatal: 'GL/gl.h' file not found
> ```
> 
> I am a little confused why there is a error message about clang

I have `llvm+clang` in my packages.yaml, so I didn't try to use this with llvm without clang before.
It failed for me with `llvm~clang`, so I added a `+clang` in py-pyside2.

OpenGL needs to be installed by the graphics driver / system, right? Or does this need a change here?
(py-pyside2 depends on qt+opengl which obviously depends on gl)

adamjstewart commented on var/spack/repos/builtin/packages/py-pyside2/package.py:
```diff
@@ -30,13 +31,15 @@ class PyPyside2(PythonPackage):
+    # dependency matrix from: https://wiki.qt.io/Qt_for_Python
+    depends_on("python@:3.10", when="@5.15.8:", type=("build", "run"))
+    depends_on("python@:3.9", when="@5.15.1:5.15.7", type=("build", "run"))
+    depends_on("python@:3.8", when="@:5.15.0", type=("build", "run"))
```
I would add:
```python
depends_on("python@:3.7", when="@:5.13", type=("build", "run"))
```
Based on PyPI wheel availability. We'll be removing 3.7 in the next release (November) so it would be good to also remove those old versions of pyside2

adamjstewart commented on var/spack/repos/builtin/packages/py-pyside2/package.py:
```diff
@@ -30,13 +31,15 @@ class PyPyside2(PythonPackage):
+    depends_on("llvm@10+clang", type="build", when="@5.15")
     depends_on("py-setuptools", type="build")
     depends_on("py-packaging", type="build")
     depends_on("py-wheel", type="build")
```
qt 5.12+ according to `sources/shiboken2/doc/gettingstarted.rst`, not sure when that changed

adamjstewart commented on var/spack/repos/builtin/packages/py-pyside2/package.py:
```diff
@@ -30,13 +31,15 @@ class PyPyside2(PythonPackage):
-    depends_on("llvm@6", type="build", when="@5.12:5.13")
-    depends_on("llvm@10", type="build", when="@5.15")
+    depends_on("llvm@6+clang", type="build", when="@5.12:5.13")
+    depends_on("llvm@10+clang", type="build", when="@5.15")
```
Can you sort from newest to oldest?

melven committed:
py-pyside2: remove old versions, update deps

adamjstewart commented on var/spack/repos/builtin/packages/py-pyside2/package.py:
```diff
@@ -17,34 +17,34 @@ class PyPyside2(PythonPackage):
+    version("5.15.10", tag="v5.15.10-lts-lgpl", submodules=True)
     version("5.15.2.1", tag="v5.15.2.1", submodules=True)
     version("5.14.2.1", tag="v5.14.2.1", submodules=True)
-    version("5.13.2", tag="v5.13.2", submodules=True)
-    version("5.13.1", tag="v5.13.1", submodules=True)
-    version("5.13.0", tag="v5.13.0", submodules=True)
-    version("5.12.5", tag="v5.12.5", submodules=True)
```
We don't normally remove versions without first deprecating them

adamjstewart commented on var/spack/repos/builtin/packages/py-pyside2/package.py:
```diff
@@ -17,34 +17,34 @@ class PyPyside2(PythonPackage):
+    depends_on("llvm@6+clang", type="build", when="@5.12:5.13")
     depends_on("py-setuptools", type="build")
     depends_on("py-packaging", type="build")
     depends_on("py-wheel", type="build")
```
wheel is added by the base class and could be removed

adamjstewart commented on var/spack/repos/builtin/packages/py-pyside2/package.py:
```diff
@@ -17,34 +17,34 @@ class PyPyside2(PythonPackage):
     # in newer pip versions --install-option does not exist
     depends_on("py-pip@:23.0", type="build")
-    depends_on("qt@5.11:+opengl")
+    depends_on("qt@5.12:+opengl")
```
qt 4.6+ for 5.12.5 according to `docs/index.rst`. You'll need to check to see what the requirements are per version. It would also be helpful to add a comment saying where these come from.

adamjstewart replied:
Could also be in `setup.py`, although that says 5.11+. Not sure which to trust

melven commented:
Outdated, I will make a new cleaned up pull request with the latest version and fixed version constraints on dependencies...
